File: Dijkstra/cpp/dijkstra_par.cpp
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>
#include <omp.h>
#include <chrono>
#include <iostream>
#include <vector>

// Compilation Instructions:
// g++ -O2 -fopenmp -std=c++11 -o dijkstraParallel dijkstraParallel.cpp

// Function to print the shortest distances from the source vertex
void printSolution(int* result, int size)
{
    printf("Vertex \t\t Distance from Source\n");
    for (int i = 0; i < size; i++)
        printf("%d \t\t\t\t %d\n", i, result[i]);
}

// Corrected parallel minDistance function
int minDistance(int* dist, bool* visited, int size)
{
    int min = INT_MAX;
    int min_index = -1;

    // Parallel region to find the minimum distance vertex
    #pragma omp parallel
    {
        int local_min = INT_MAX;
        int local_min_index = -1;

        #pragma omp for nowait
        for (int i = 0; i < size; i++) {
            if (!visited[i] && dist[i] < local_min) {
                local_min = dist[i];
                local_min_index = i;
            }
        }

        // Critical section to update the global minimum
        #pragma omp critical
        {
            if (local_min < min) {
                min = local_min;
                min_index = local_min_index;
            }
        }
    }

    return min_index;
}

// Parallel Dijkstra's algorithm
void dijkstra(int* graph, int src, int size){
    // Allocate memory for distances and visited arrays
    int* distances = (int*)malloc(size * sizeof(int));
    bool* visited = (bool*)malloc(size * sizeof(bool));

    if (distances == NULL || visited == NULL) {
        std::cerr << "Memory allocation failed." << std::endl;
        exit(EXIT_FAILURE);
    }

    // Initialize distances and visited arrays in parallel
    #pragma omp parallel for
    for (int i = 0; i < size; i++) {
        distances[i] = INT_MAX;
        visited[i] = false;
    }

    distances[src] = 0;

    for (int count = 0; count < size - 1; count++) {
        // Pick the minimum distance vertex from the set of vertices not yet processed
        int u = minDistance(distances, visited, size);

        // If the smallest distance is INT_MAX, remaining vertices are inaccessible
        if (u == -1 || distances[u] == INT_MAX)
            break;

        // Mark the picked vertex as processed
        visited[u] = true;

        // Update distances of adjacent vertices in parallel
        #pragma omp parallel for
        for (int v = 0; v < size; v++) {
            if (!visited[v] && graph[u*size + v] && 
                distances[u] != INT_MAX && 
                distances[u] + graph[u*size + v] < distances[v]) {
                distances[v] = distances[u] + graph[u*size + v];
            }
        }
    }

    // Uncomment the following line to print the shortest distances
    // printSolution(distances, size);

    // Free allocated memory
    free(distances);
    free(visited);
}

// Function to generate a random adjacency matrix for an undirected graph
void generateAdjMatrix(int* adjMatrix, int size){
    int temp;

    // Seed the random number generator
    srand(time(NULL));

    #pragma omp parallel for private(temp)
    for(int i = 0; i < size; i++){
        for(int j = i; j < size; j++){
            if(i == j) {
                adjMatrix[i*size+j] = 0;
                continue;
            }
            temp = rand() % 10;
            adjMatrix[i*size + j] = temp;
            adjMatrix[j*size + i] = temp;
        }
    }

    // Optional: Print the adjacency matrix
    /*
    for(int i = 0; i < size; i++){
        for(int j = 0; j < size; j++){
            // printf("%d ", adjMatrix[i*size + j]);
        }
        printf("\n");
    }
    */
}

int main(int argc, char* argv[]){

    if (argc < 2 || argc > 3) {
        std::cerr << "Usage: " << argv[0] << " <graph_size> [num_threads]" << std::endl;
        return EXIT_FAILURE;
    }

    int size = atoi(argv[1]);

    if (size <= 0) {
        std::cerr << "Error: Graph size must be a positive integer." << std::endl;
        return EXIT_FAILURE;
    }

    // Determine the number of threads
    int num_threads;
    if (argc == 3) {
        num_threads = atoi(argv[2]);
        if (num_threads <= 0) {
            std::cerr << "Error: Number of threads must be a positive integer." << std::endl;
            return EXIT_FAILURE;
        }
        omp_set_num_threads(num_threads);
    } else {
        // Default to maximum available threads
        num_threads = omp_get_max_threads();
    }

    // Allocate memory for the adjacency matrix
    int* graph = (int*)calloc(size*size, sizeof(int));
    if (graph == NULL) {
        std::cerr << "Memory allocation for graph failed." << std::endl;
        return EXIT_FAILURE;
    }

    // Generate the adjacency matrix
    generateAdjMatrix(graph, size);

    std::cout << "Running parallel Dijkstra's algorithm with graph size " 
              << size << "x" << size << " using " << num_threads << " thread(s)..." << std::endl;

    // Start timing
    auto start = std::chrono::high_resolution_clock::now();

    // Execute Dijkstra's algorithm
    dijkstra(graph, 0, size);

    // End timing
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::micro> duration = end - start;
    double seconds = duration.count() / 1000000.0;

    std::cout << "Execution Time: " << duration.count() << " microseconds" << std::endl;

    // Uncomment the following line to print the shortest distances
    // printSolution(distances, size);

    // Free allocated memory for the graph
    free(graph);

    return EXIT_SUCCESS;
}

File: Dijkstra/rust/src/bin/dijkstra_par.rs
use std::env;
use rand::Rng;
use rayon::prelude::*;
use std::time::Instant;

/// Finds the vertex with the minimum distance value from the set of vertices not yet processed.
/// Utilizes Rayon’s parallel iterators for efficient computation.
fn min_distance(distances: &Vec<i32>, visited: &Vec<bool>) -> Option<usize> {
    distances.par_iter()
        .enumerate()
        .filter(|(i, _)| !visited[*i])
        .min_by_key(|&(_, dist)| dist)
        .map(|(i, _)| i)
}

/// Dijkstra's algorithm to find the shortest path from a start vertex.
/// Optimized to leverage Rayon for parallel computation of distance updates.
fn dijkstra(graph: &Vec<Vec<Option<usize>>>, start: usize, size: usize) -> Vec<i32> {
    // Initialize distances as infinity and visited as false for all vertices
    let mut distances = vec![i32::MAX; size];
    let mut visited = vec![false; size];

    distances[start] = 0; // Distance to the start vertex is 0

    for _ in 0..size - 1 {
        // Select the unvisited vertex with the smallest distance
        let u = match min_distance(&distances, &visited) {
            Some(u) => u,
            None => break, // All reachable vertices have been processed
        };

        if distances[u] == i32::MAX {
            break; // Remaining vertices are inaccessible from the source
        }

        visited[u] = true; // Mark the vertex as visited

        // Collect potential distance updates in parallel
        let new_distances: Vec<(usize, i32)> = graph[u].par_iter()
            .enumerate()
            .filter_map(|(v, &weight)| {
                if let Some(w) = weight {
                    if !visited[v] && distances[u] != i32::MAX {
                        let next_distance = distances[u] + w as i32;
                        if next_distance < distances[v] {
                            Some((v, next_distance))
                        } else {
                            None
                        }
                    } else {
                        None
                    }
                } else {
                    None
                }
            })
            .collect();

        // Apply the collected distance updates
        for (v, dist) in new_distances {
            distances[v] = dist;
        }
    }

    distances
}

/// Generates a random adjacency matrix for an undirected graph.
/// Each edge has a 70% chance of existing with a random weight between 1 and `max_weight`.
fn generate_graph(num_vertices: usize, max_weight: usize) -> Vec<Vec<Option<usize>>> {
    let mut rng = rand::thread_rng();
    let mut graph: Vec<Vec<Option<usize>>> = vec![vec![None; num_vertices]; num_vertices];

    for i in 0..num_vertices {
        for j in 0..num_vertices {
            if i != j && rng.gen_bool(0.7) {
                graph[i][j] = Some(rng.gen_range(1..=max_weight));
            }
        }
    }

    graph
}

fn main() {
    // Parse command-line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 || args.len() > 3 {
        eprintln!("Usage: {} <graph_size> [num_threads]", args[0]);
        std::process::exit(1);
    }

    // Parse the graph size
    let size: usize = match args[1].parse() {
        Ok(n) if n > 0 => n,
        _ => {
            eprintln!("Error: <graph_size> must be a positive integer.");
            std::process::exit(1);
        }
    };

    // Optionally parse the number of threads (currently unused as Rayon manages the thread pool)
    if args.len() == 3 {
        let num_threads: usize = match args[2].parse() {
            Ok(n) if n > 0 => n,
            _ => {
                eprintln!("Error: [num_threads] must be a positive integer.");
                std::process::exit(1);
            }
        };
        rayon::ThreadPoolBuilder::new()
            .num_threads(num_threads)
            .build_global()
            .unwrap_or_else(|_| {
                eprintln!("Error: Failed to build the Rayon thread pool.");
                std::process::exit(1);
            });
    }

    let max_weight = 20;
    let start_vertex = 0;

    // Generate a random graph
    let graph = generate_graph(size, max_weight);

    // Uncomment the following lines to print the generated graph
    /*
    println!("Generated Graph (Adjacency Matrix):");
    for row in &graph {
        for &weight in row {
            match weight {
                Some(w) => print!("{:3} ", w),
                None => print!("  . "),
            }
        }
        println!();
    }
    */

    println!(
        "Running parallel Dijkstra's algorithm with graph size {}x{} using Rayon...",
        size, size
    );

    // Start timing
    let start = Instant::now();

    // Execute Dijkstra's algorithm
    let distances = dijkstra(&graph, start_vertex, size);

    // End timing
    let duration = start.elapsed();
    println!("Time taken for parallel Dijkstra in Rust for size {1} by {1}: {0:?}", duration, size);
    

    // Uncomment the following lines to print the shortest distances
    /*
    println!("\nShortest distances from vertex {}:", start_vertex);
    for (i, &dist) in distances.iter().enumerate() {
        println!(
            "Vertex {}: {}",
            i,
            if dist == i32::MAX {
                "∞".to_string()
            } else {
                dist.to_string()
            }
        );
    }
    */
}
